File: files/activities.ts
/**
 * Temporal Activities for Claude Code Integration
 * These are the actual operations that interact with Claude Code CLI
 */

import { exec } from 'child_process';
import { promisify } from 'util';
import * as fs from 'fs/promises';
import * as path from 'path';

const execAsync = promisify(exec);

interface ClaudeCodeResult {
  output: string;
  tokensUsed: number;
  cost: number;
  duration: number;
  filesModified: string[];
  linesAdded: number;
  linesRemoved: number;
  diffUrl?: string;
}

interface CostEstimate {
  estimated: number;
  model: string;
  tokensEstimate: number;
}

interface MetricsData {
  [key: string]: any;
  timestamp?: string;
}

/**
 * Execute Claude Code with detailed telemetry
 * 
 * ADVANTAGE: Captures all execution metadata automatically
 */
export async function executeClaudeCode(params: {
  prompt: string;
  workingDirectory: string;
  maxTokens?: number;
  temperature?: number;
}): Promise<ClaudeCodeResult> {
  
  const startTime = Date.now();
  const { prompt, workingDirectory, maxTokens = 8000, temperature = 0.3 } = params;

  console.log(`\n[Activity] Executing Claude Code...`);
  console.log(`Directory: ${workingDirectory}`);
  console.log(`Prompt: ${prompt.substring(0, 100)}...`);

  try {
    // Store git state before changes
    const beforeFiles = await getGitStatus(workingDirectory);

    // Execute Claude Code
    // In production, use actual Claude Code CLI
    const command = `cd ${workingDirectory} && echo "${prompt}" | claude --json`;
    
    const { stdout, stderr } = await execAsync(command, {
      env: {
        ...process.env,
        CLAUDE_MAX_TOKENS: maxTokens.toString(),
        CLAUDE_TEMPERATURE: temperature.toString(),
      },
      maxBuffer: 10 * 1024 * 1024, // 10MB buffer
    });

    // Parse Claude Code output (mock structure for demo)
    const output = stdout || stderr;
    
    // Get git changes after execution
    const afterFiles = await getGitStatus(workingDirectory);
    const filesModified = getChangedFiles(beforeFiles, afterFiles);

    // Calculate git stats
    const gitDiff = await getGitDiff(workingDirectory);
    const stats = parseGitStats(gitDiff);

    // Estimate tokens and cost (in production, parse from Claude Code output)
    const tokensUsed = estimateTokens(prompt + output);
    const cost = calculateCost(tokensUsed, 'claude-sonnet-4-5');

    const duration = Date.now() - startTime;

    const result: ClaudeCodeResult = {
      output,
      tokensUsed,
      cost,
      duration,
      filesModified,
      linesAdded: stats.additions,
      linesRemoved: stats.deletions,
    };

    console.log(`[Activity] Completed in ${duration}ms`);
    console.log(`Tokens: ${tokensUsed}, Cost: $${cost.toFixed(4)}`);
    console.log(`Files modified: ${filesModified.length}`);

    return result;

  } catch (error) {
    console.error(`[Activity] Failed:`, error.message);
    throw new Error(`Claude Code execution failed: ${error.message}`);
  }
}

/**
 * Run test suite with detailed results
 * 
 * ADVANTAGE: Automatic validation after each change
 */
export async function runTests(projectPath: string): Promise<{
  success: boolean;
  totalTests: number;
  passed: number;
  failed: number;
  duration: number;
  errors: string[];
  coverage?: number;
}> {
  
  console.log(`\n[Activity] Running tests in ${projectPath}...`);
  const startTime = Date.now();

  try {
    // Run npm test (adjust for your test framework)
    const { stdout, stderr } = await execAsync(
      'npm test -- --json --coverage',
      { cwd: projectPath }
    );

    const duration = Date.now() - startTime;

    // Parse test results (adjust parsing based on your test framework)
    const result = parseTestResults(stdout);

    console.log(`[Activity] Tests completed in ${duration}ms`);
    console.log(`Passed: ${result.passed}/${result.totalTests}`);

    return {
      ...result,
      duration,
      success: result.failed === 0,
    };

  } catch (error) {
    // Tests failed
    return {
      success: false,
      totalTests: 0,
      passed: 0,
      failed: 1,
      duration: Date.now() - startTime,
      errors: [error.message],
    };
  }
}

/**
 * Estimate cost before execution
 * 
 * ADVANTAGE: Budget control and cost awareness
 */
export async function estimateCost(params: {
  prompt: string;
  complexity: 'low' | 'medium' | 'high';
}): Promise<CostEstimate> {
  
  const { prompt, complexity } = params;
  
  // Base token estimate from prompt
  const promptTokens = estimateTokens(prompt);
  
  // Estimate completion tokens based on complexity
  const completionMultiplier = {
    low: 2,
    medium: 4,
    high: 8,
  };
  
  const estimatedCompletionTokens = promptTokens * completionMultiplier[complexity];
  const totalTokens = promptTokens + estimatedCompletionTokens;
  
  const cost = calculateCost(totalTokens, 'claude-sonnet-4-5');
  
  return {
    estimated: cost,
    model: 'claude-sonnet-4-5',
    tokensEstimate: totalTokens,
  };
}

/**
 * Create project snapshot for rollback capability
 * 
 * ADVANTAGE: Safe experimentation with instant rollback
 */
export async function createSnapshot(projectPath: string): Promise<string> {
  
  const snapshotId = `snapshot-${Date.now()}`;
  console.log(`\n[Activity] Creating snapshot: ${snapshotId}`);

  try {
    // Create git stash or commit
    await execAsync('git add -A', { cwd: projectPath });
    await execAsync(`git commit -m "Snapshot: ${snapshotId}" --allow-empty`, {
      cwd: projectPath,
    });

    console.log(`[Activity] Snapshot created: ${snapshotId}`);
    return snapshotId;

  } catch (error) {
    console.warn(`Failed to create snapshot: ${error.message}`);
    return snapshotId; // Return ID anyway for tracking
  }
}

/**
 * Notify developer with context
 * 
 * ADVANTAGE: Human-in-the-loop with rich context
 */
export async function notifyDeveloper(params: {
  stage: string;
  message: string;
  filesChanged: string[];
  diffUrl?: string;
}): Promise<void> {
  
  console.log(`\n[Activity] Notifying developer...`);
  console.log(`Stage: ${params.stage}`);
  console.log(`Message: ${params.message}`);
  console.log(`Files changed: ${params.filesChanged.join(', ')}`);
  
  // In production, integrate with:
  // - Slack API
  // - Email service
  // - GitHub notifications
  // - Custom webhook
  
  // Mock notification
  await new Promise(resolve => setTimeout(resolve, 100));
  
  console.log(`[Activity] Notification sent`);
}

/**
 * Capture metrics to monitoring system
 * 
 * ADVANTAGE: Complete observability and analytics
 */
export async function captureMetrics(data: MetricsData): Promise<void> {
  
  const metrics = {
    ...data,
    timestamp: new Date().toISOString(),
  };

  console.log(`\n[Activity] Capturing metrics:`, metrics);

  // In production, send to:
  // - Prometheus Pushgateway
  // - Datadog
  // - CloudWatch
  // - Custom metrics endpoint

  // Mock metrics storage
  const metricsFile = '/tmp/claude-code-metrics.jsonl';
  await fs.appendFile(
    metricsFile,
    JSON.stringify(metrics) + '\n',
    'utf-8'
  );

  console.log(`[Activity] Metrics captured`);
}

/**
 * Deploy to NPM (example integration)
 * 
 * ADVANTAGE: End-to-end automation
 */
export async function deployToNPM(params: {
  projectPath: string;
  version: string;
  tag?: string;
}): Promise<void> {
  
  console.log(`\n[Activity] Deploying to NPM...`);
  console.log(`Version: ${params.version}`);

  const { projectPath, version, tag = 'latest' } = params;

  try {
    // Update version in package.json
    await execAsync(`npm version ${version} --no-git-tag-version`, {
      cwd: projectPath,
    });

    // Publish to NPM
    await execAsync(`npm publish --tag ${tag}`, {
      cwd: projectPath,
    });

    console.log(`[Activity] Successfully published ${version} to NPM`);

  } catch (error) {
    throw new Error(`NPM deployment failed: ${error.message}`);
  }
}

// Helper functions

async function getGitStatus(cwd: string): Promise<string> {
  try {
    const { stdout } = await execAsync('git status --short', { cwd });
    return stdout;
  } catch {
    return '';
  }
}

async function getGitDiff(cwd: string): Promise<string> {
  try {
    const { stdout } = await execAsync('git diff --stat', { cwd });
    return stdout;
  } catch {
    return '';
  }
}

function getChangedFiles(before: string, after: string): string[] {
  const beforeFiles = new Set(before.split('\n').map(line => line.split(' ').pop()));
  const afterFiles = after.split('\n').map(line => line.split(' ').pop());
  return afterFiles.filter(f => f && !beforeFiles.has(f));
}

function parseGitStats(diff: string): { additions: number; deletions: number } {
  // Parse git diff --stat output
  const match = diff.match(/(\d+) insertions?.*?(\d+) deletions?/);
  return {
    additions: match ? parseInt(match[1]) : 0,
    deletions: match ? parseInt(match[2]) : 0,
  };
}

function estimateTokens(text: string): number {
  // Rough estimate: ~4 characters per token
  return Math.ceil(text.length / 4);
}

function calculateCost(tokens: number, model: string): number {
  // Pricing as of 2025 (adjust as needed)
  const pricing: Record<string, { input: number; output: number }> = {
    'claude-sonnet-4-5': { input: 0.003, output: 0.015 }, // per 1K tokens
    'claude-opus-4': { input: 0.015, output: 0.075 },
  };

  const rates = pricing[model] || pricing['claude-sonnet-4-5'];
  // Assume 50/50 split input/output for simplicity
  const cost = (tokens / 1000) * ((rates.input + rates.output) / 2);
  return cost;
}

function parseTestResults(output: string): {
  totalTests: number;
  passed: number;
  failed: number;
  errors: string[];
} {
  // Mock parsing - adjust based on your test framework
  try {
    const json = JSON.parse(output);
    return {
      totalTests: json.numTotalTests || 0,
      passed: json.numPassedTests || 0,
      failed: json.numFailedTests || 0,
      errors: json.testResults?.map((t: any) => t.message) || [],
    };
  } catch {
    return {
      totalTests: 0,
      passed: 0,
      failed: 0,
      errors: ['Failed to parse test results'],
    };
  }
}

File: claude_temporal/activities.py
"""
Temporal Activities for Claude Code Integration.

These are the actual operations that interact with Claude Code CLI.
Each activity captures telemetry for observability.
"""

import asyncio
import json
import os
import time
from pathlib import Path

from temporalio import activity

from .models import (
    ClaudeCodeInput,
    ClaudeCodeResult,
    TestResult,
    CostEstimate,
    NotificationParams,
    MetricsData,
)
from .config import get_config
from .constants import (
    CHARS_PER_TOKEN,
    COMPLEXITY_MULTIPLIERS,
    MODEL_PRICING,
    DEFAULT_MODEL,
)
from .git_utils import GitOperations
from .notification import get_notification_service
from .test_runner import get_test_runner


def estimate_tokens(text: str) -> int:
    """Estimate token count from text."""
    return len(text) // CHARS_PER_TOKEN


def calculate_cost(tokens: int, model: str = DEFAULT_MODEL) -> float:
    """Calculate cost based on token usage and model pricing."""
    rates = MODEL_PRICING.get(model, MODEL_PRICING[DEFAULT_MODEL])
    # Assume 50/50 split input/output for simplicity
    return (tokens / 1000) * ((rates["input"] + rates["output"]) / 2)


def _validate_path(path: str) -> Path:
    """
    Validate that a path exists.

    Args:
        path: Path to validate

    Returns:
        Path object

    Raises:
        ValueError: If path doesn't exist
    """
    p = Path(path)
    if not p.exists():
        raise ValueError(f"Path does not exist: {path}")
    return p


@activity.defn
async def execute_claude_code(params: ClaudeCodeInput) -> ClaudeCodeResult:
    """
    Execute Claude Code with detailed telemetry.

    This activity runs the Claude CLI and captures all execution metadata
    for observability and cost tracking.
    """
    start_time = time.time()
    config = get_config()

    activity.logger.info(f"Executing Claude Code in {params.working_directory}")
    activity.logger.info(f"Prompt: {params.prompt[:100]}...")

    # Validate working directory
    _validate_path(params.working_directory)

    # Initialize git operations
    git = GitOperations(params.working_directory)

    # Store git state before changes
    before_status = await git.get_status()

    try:
        # Execute Claude Code CLI
        # Using --print flag for non-interactive output
        proc = await asyncio.create_subprocess_exec(
            "claude",
            "--print",
            params.prompt,
            cwd=params.working_directory,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
            env={
                **os.environ,
                "CLAUDE_MAX_TOKENS": str(params.max_tokens),
            },
        )

        stdout, stderr = await asyncio.wait_for(
            proc.communicate(),
            timeout=config.claude.timeout_seconds,
        )

        output = stdout.decode() if stdout else stderr.decode()

        # Get git changes after execution
        after_status = await git.get_status()
        files_modified = after_status.files_changed

        # Get diff stats
        diff_stats = await git.get_diff_stats()

        # Estimate tokens and cost
        tokens_used = estimate_tokens(params.prompt + output)
        cost = calculate_cost(tokens_used)

        duration_ms = int((time.time() - start_time) * 1000)

        result = ClaudeCodeResult(
            output=output,
            tokens_used=tokens_used,
            cost=cost,
            duration_ms=duration_ms,
            files_modified=files_modified,
            lines_added=diff_stats.lines_added,
            lines_removed=diff_stats.lines_removed,
        )

        activity.logger.info(f"Completed in {duration_ms}ms")
        activity.logger.info(f"Tokens: {tokens_used}, Cost: ${cost:.4f}")
        activity.logger.info(f"Files modified: {len(files_modified)}")

        return result

    except asyncio.TimeoutError:
        raise RuntimeError(
            f"Claude Code execution timed out after {config.claude.timeout_seconds}s"
        )
    except Exception as e:
        activity.logger.error(f"Failed: {e}")
        raise RuntimeError(f"Claude Code execution failed: {e}")


@activity.defn
async def run_tests(project_path: str) -> TestResult:
    """
    Run test suite with detailed results.

    Automatically detects and uses the appropriate test framework.
    """
    activity.logger.info(f"Running tests in {project_path}")

    # Validate path
    _validate_path(project_path)

    # Get auto-detecting test runner
    runner = get_test_runner()

    # Run tests
    result = await runner.run(project_path)

    activity.logger.info(f"Tests completed in {result.duration_ms}ms")
    activity.logger.info(f"Passed: {result.passed}/{result.total_tests}")

    return result


@activity.defn
async def estimate_cost(prompt: str, complexity: str = "medium") -> CostEstimate:
    """
    Estimate cost before execution for budget control.
    """
    # Base token estimate from prompt
    prompt_tokens = estimate_tokens(prompt)

    # Estimate completion tokens based on complexity
    multiplier = COMPLEXITY_MULTIPLIERS.get(complexity, COMPLEXITY_MULTIPLIERS["medium"])
    completion_tokens = prompt_tokens * multiplier

    total_tokens = prompt_tokens + completion_tokens
    cost = calculate_cost(total_tokens)

    return CostEstimate(
        estimated=cost,
        model=DEFAULT_MODEL,
        tokens_estimate=total_tokens,
    )


@activity.defn
async def create_snapshot(project_path: str) -> str:
    """
    Create project snapshot for rollback capability.

    Uses git commits as snapshots for safe experimentation.
    """
    snapshot_id = f"snapshot-{int(time.time())}"
    activity.logger.info(f"Creating snapshot: {snapshot_id}")

    try:
        # Validate path
        _validate_path(project_path)

        git = GitOperations(project_path)
        result = await git.create_snapshot(snapshot_id)

        if result.success:
            activity.logger.info(f"Snapshot created: {snapshot_id}")
        else:
            activity.logger.warning(f"Snapshot creation issue: {result.error}")

        return snapshot_id

    except Exception as e:
        activity.logger.warning(f"Failed to create snapshot: {e}")
        return snapshot_id  # Return ID anyway for tracking


@activity.defn
async def notify_developer(params: NotificationParams) -> None:
    """
    Notify developer with context for human-in-the-loop approval.

    Uses configurable notification service (console, Slack, webhook, etc.)
    """
    activity.logger.info(f"Notifying developer...")
    activity.logger.info(f"Stage: {params.stage}")
    activity.logger.info(f"Message: {params.message}")
    activity.logger.info(f"Files changed: {', '.join(params.files_changed)}")

    # Get notification service from config
    config = get_config()
    notification_config = {
        "type": config.notification.type,
        "webhook_url": config.notification.slack_webhook_url,
        "channel": config.notification.slack_channel,
        "url": config.notification.webhook_url,
        "headers": config.notification.webhook_headers,
    }

    service = get_notification_service(notification_config)
    success = await service.send(params)

    if success:
        activity.logger.info(f"Notification sent via {service.get_name()}")
    else:
        activity.logger.warning(f"Failed to send notification via {service.get_name()}")


@activity.defn
async def capture_metrics(data: MetricsData) -> None:
    """
    Capture metrics to monitoring system for complete observability.

    Writes metrics to file for later analysis.
    """
    config = get_config()

    metrics = {
        "stage": data.stage,
        "tokens_used": data.tokens_used,
        "cost": data.cost,
        "duration_ms": data.duration_ms,
        "files_modified": data.files_modified,
        "lines_added": data.lines_added,
        "lines_removed": data.lines_removed,
        "tests_pass": data.tests_pass,
        "error": data.error,
        "timestamp": data.timestamp,
    }

    activity.logger.info(f"Capturing metrics: {metrics}")

    # Write to metrics file asynchronously
    try:
        metrics_path = Path(config.metrics_file)
        metrics_path.parent.mkdir(parents=True, exist_ok=True)

        # Use async file writing via run_in_executor
        loop = asyncio.get_event_loop()
        await loop.run_in_executor(
            None,
            _write_metrics,
            metrics_path,
            metrics,
        )

        activity.logger.info("Metrics captured")
    except Exception as e:
        activity.logger.warning(f"Failed to write metrics: {e}")


def _write_metrics(path: Path, metrics: dict) -> None:
    """Synchronous helper for writing metrics to file."""
    with open(path, "a") as f:
        f.write(json.dumps(metrics) + "\n")


@activity.defn
async def restore_snapshot(project_path: str, snapshot_id: str) -> bool:
    """
    Restore project to a previous snapshot.
    """
    activity.logger.info(f"Restoring snapshot: {snapshot_id}")

    try:
        # Validate path
        _validate_path(project_path)

        git = GitOperations(project_path)
        result = await git.restore_snapshot(snapshot_id)

        if result.success:
            activity.logger.info(f"Restored to snapshot: {snapshot_id}")
            return True
        else:
            activity.logger.error(f"Failed to restore: {result.error}")
            return False

    except Exception as e:
        activity.logger.error(f"Failed to restore snapshot: {e}")
        return False
